core/src/main/java/org/apache/calcite/rel/metadata/FunctionalDependencySet.java

```diff
@@ -0,0 +1,420 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.metadata;
+
+import org.apache.calcite.util.ImmutableBitSet;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A set of functional dependencies with closure and minimal cover operations.
+ * This class implements standard algorithms for functional dependency reasoning.
+ */
+public class FunctionalDependencySet {
+  // Maximum number of transitive closure iterations to prevent infinite loops
+  public static final int MAX_TRANSITIVE_CLOSURE_LOOPS = 10;
+  // Maximum number of attributes supported in closure computation
+  private static final int MAX_CLOSURE_ATTRS = 10000;
+
+  private final Set<FunctionalDependency> fdSet = new HashSet<>();
+
+  public FunctionalDependencySet() {}
+
+  public FunctionalDependencySet(Set<FunctionalDependency> fds) {
+    this.fdSet.addAll(fds);
+  }
+
+  public void addFD(FunctionalDependency fd) {
```

Comment:
can somehow the functions that mutate `this` be documented?
Some operations, like `union` are functional, others, like `add` are not.
(it's true that add returns void, but this doesn't preclude having a variant of `add` that returns `this`.)